README
NAME
    Music::Tag - Interface for collecting information about music files.

DESCRIPTION
    Extendable module for working with Music Tags. Music::Tag Is powered by
    various plugins that collect data about a song based on whatever
    information has already been discovered.

    The motivation behind this was to provide a convenient method for fixing
    broken tags in music files. This developed into a universal interface to
    various music file tagging schemes and a convenient way to augment this
    from online databases.

    Several plugin modules to find information about a music file and write
    it back into the tag are available. These modules will use available
    information (REQUIRED DATA VALUES and USED DATA VALUES) and set various
    data values back to the tag.

INSTALLATION
    To install this module type the following:

       perl Makefile.PL
       make
       make test
       make install

  IMPORTANT NOTE
    If you have installed older versions (older than .25) PLEASE delete the
    following scripts from your bin folder: autotag, safetag, quicktag,
    musicsort, musicinfo.

    If you used any of these scripts, create a symbolic link to musictag for
    each.

  QUICK INSTALL OF ALL PACKAGES
    A bundle is available to quickly install Music::Tag with all plugins. To
    install it use:

       perl -MCPAN -eshell

    At the cpan shell prompt type:

       install Bundle::Music::Tag

DEPENDENCIES
    This module requires these other modules and libraries:

       Encode
       File::Spec
       Locale::Country
       Digest::SHA1
       Config::Options

    I strongly recommend the following to improve web searches:

       Lingua::EN::Inflect
       Text::LevenshteinXS
       Text::Unaccent::PurePerl

    The following just makes things pretty:

       Term::ANSIColor

EXECUTABLE SCRIPT
    An executable script, musictag is allows quick tagging of MP3 files. To
    learn more, use:

       musictag --help 
       musictag --longhelp

SOURCE
    Source is available at github: http://github.com/riemann42/Music-Tag.

BUG TRACKING
    Please use github for bug tracking:
    http://github.com/riemann42/Music-Tag/issues.

AUTHOR
    Edward Allen III <ealleniii _at_ cpan _dot_ org>

COPYRIGHT
    Copyright (c) 2007,2008,2010 Edward Allen III. Some rights reserved.

LICENSE
    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself, either:

    a) the GNU General Public License as published by the Free Software
    Foundation; either version 1, or (at your option) any later version, or

    b) the "Artistic License" which comes with Perl.

    This program is distributed in the hope that it will be useful, but
    WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See either the GNU
    General Public License or the Artistic License for more details.

    You should have received a copy of the Artistic License with this Kit,
    in the file named "Artistic". If not, I'll be glad to provide one.

    You should also have received a copy of the GNU General Public License
    along with this program in the file named "Copying". If not, write to
    the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
    Boston, MA 02110-1301, USA or visit their web page on the Internet at
    http://www.gnu.org/copyleft/gpl.html.



CHANGES
	Release Name: 0.40_03
		*	Text::Levenshtien was not being detected.

	Release Name: 0.40_02
		*	Fixed minor testing issue, which caused test to assume that
			Text::Levenshtien was installed.

	Release Name: 0.40_01 ( Normalized Version number accross plugins)
		*	Started using git and github

		* 	Added capability methods to plugins (required_values, set_values
			and saved_values).  Not sure what to do with them now...

		*	Significant improvment in testing, which found bugs that
			were fixed.

			* Tracknum and Discnum didn't work correctly

			* Epoch times failed completly.

		*	Fixed severe memory leak by with weaken

		*	Changed accessor modules to not use Autoload

		*	Fixed a few open file handles that didn't have a close.

    Release Name: 0.33
        *   Revised POD (thanks Ivan Tubert-Brohman for Test::Spelling!)

        *   Added the ability for plugins to set a verbosity level with
            status method

        *   Added datamethods upc and ean with value checking. Sets one if
            other is set.

        *   Cleaned up some of the code

        *   Started using Pod::Readme for README and CHANGES

    Release Name: 0.32
        *   Fixed critical bug with Term::ANSIColor

    Release Name: 0.31
        *   Added example

        *   Added used_datamethods method

    Release Name: 0.30
        *   Config::Options .07 now required

        *   POD and Kwalitee changes

    Release Name: 0.29
        *   Fixed bug in autoplugin default settings preventing ogg and flac
            from working.

    Release Name: 0.28
        *   Seperated plugins into seperate distributions

        *   Revised module detection code to actually work

        *   Plugins for autoplugin process are now loaded based on a new
            option, autoplugin This option is a hash ref of file extenstions
            to plugins

        *   Revised POD

        *   Added simple test script now that plugins have been seperated

        *   Revised help in musictag script

        *   Added option to use --[presetname] to call a preset in musictag

    Release Name: 0.27
        *   More documentation and tested POD.

        *   datamethods method now can be used to add new datamethods

        *   Added test for MusicBrainz and Amazon plugins

        *   Revised releasedate and recorddate internal storage to store as
            releasetime and recordtime -- with full timestamps

        *   Added releasetime, recordtime, releaseepoch, and recordepoech
            datamethods

        *   Support for TIME ID3v2 tag

        *   After much thought, replaced Ogg::Vorbis::Header with
            Ogg::Vorbis::Header::PurePerl and added vorbiscomment to write
            tags

        *   Revised OGG and FLAC plugins to clean up code (much slicker now)

    Release Name: 0.26
        *   Removed several prerequistes that weren't used

        *   Fixed error in README about prerequisite

    Release Name: 0.25
        *   Support many more tags for flac, ogg, and m4a

        *   Removed autotag safetag quicktag musictag musicsort musicinfo
            scripts All is done by musictag now

        *   Added tests for some plugins. More to do!

        *   Bug Fixes

        *   Documentation improvments

        *   Added preset option for musictag

    Release Name: 0.24
        *   Bug Fixes

        *   Revised MP3 Tags to read Picard tags

    Release Name: 0.23
        *   Initial Public Release



lib/Music/Tag/Test.pm
package Music::Tag::Test;
use base 'Exporter';
use vars '@EXPORT';
use strict;
use Test::More;
use Digest::SHA1;
use File::Copy;
use 5.006;

@EXPORT = qw(create_tag read_tag random_write random_read random_write_num random_read_num random_write_date random_read_date filetest);

my %values = ();

sub create_tag {
    my $filetest    = shift;
    my $tagoptions  = shift;
	my $testoptions = shift;
	return 0 unless (-f $filetest);
	my $tag = Music::Tag->new($filetest, $tagoptions, $testoptions->{plugin} || 'Auto');
	ok($tag, 'Object created: ' . $filetest);
	die unless $tag;
	ok($tag->get_tag,           'get_tag called: ' . $filetest);
	ok($tag->isa('Music::Tag'), 'Correct Class: ' . $filetest);
	return $tag;
}

sub read_tag {
	my $tag = shift;
	my $testoptions = shift;
	return 0 if (! exists $testoptions->{values_in});
	my $c=0;
	foreach my $meth (keys %{$testoptions->{values_in}}) {
		SKIP: {
			skip "$meth test skipped", 1 if (! $testoptions->{values_in}->{$meth});
			$c++;
			cmp_ok($tag->$meth, 'eq', $testoptions->{values_in}->{$meth});
		}
	}
	return $c;
}

sub random_write {
	my $tag = shift;
	my $testoptions = shift;
	return 0 if (! exists $testoptions->{random_write});
	my $c = 0;
	foreach my $meth (@{$testoptions->{random_write}}) {
		my $val = "test" . $meth . int(rand(1000));
		$values{$meth} = $val;
		ok($tag->$meth($val), 'auto write to ' . $meth);
		$c++;
	}
	return $c;
}

sub random_write_num {
	my $tag = shift;
	my $testoptions = shift;
	return 0 if (! exists $testoptions->{random_write_num});
	my $c = 0;
	foreach my $meth (@{$testoptions->{random_write_num}}) {
		my $val = int(rand(10))+1;
		$values{$meth} = $val;
		ok($tag->$meth($val), 'auto write to ' . $meth);
		$c++;
	}
	return $c;
}

sub random_write_date {
	my $tag = shift;
	my $testoptions = shift;
	return 0 if (! exists $testoptions->{random_write_date});
	my $c = 0;
	foreach my $meth (@{$testoptions->{random_write_date}}) {
		 my $val = int(rand(1_800_000_000));
		 $values{$meth} = $val;
		 ok($tag->$meth($val), 'auto write to '. $meth);
		 $c++;
	}
	return $c;
}

sub random_read {
	my $tag = shift;
	my $testoptions = shift;
	return 0 if (! exists $testoptions->{random_write});
	my $c = 0;
	foreach my $meth (@{$testoptions->{random_write}}) {
		cmp_ok($tag->$meth, 'eq', $values{$meth}, 'auto read of ' . $meth);
		$c++;
	}
	return $c;
}

sub random_read_num {
	my $tag = shift;
	my $testoptions = shift;
	return 0 if (! exists $testoptions->{random_write_num});
	my $c = 0;
	foreach my $meth (@{$testoptions->{random_write_num}}) {
		cmp_ok($tag->$meth, '==', $values{$meth}, 'auto read of ' . $meth);
		$c++;
	}
	return $c;
}

sub random_read_date {
	my $tag = shift;
	my $testoptions = shift;
	return 0 if (! exists $testoptions->{random_write_date});
	my $c = 0;
	foreach my $meth (@{$testoptions->{random_write_date}}) {
		 my $meth_t = $meth;
		 $meth_t =~ s/epoch/time/;
		 my $meth_d = $meth;
		 $meth_d =~ s/epoch/date/;
		 $meth_d =~ s/_date//;
		 my @tm = gmtime($values{$meth});
		 cmp_ok(substr($tag->$meth_t,0,16), 'eq', substr(sprintf('%04d-%02d-%02d %02d:%02d:%02d', $tm[5]+1900, $tm[4]+1, $tm[3], $tm[2], $tm[1], $tm[0]),0,16), 'auto read from '. $meth_t);
		 cmp_ok($tag->$meth_d, 'eq', sprintf('%04d-%02d-%02d', $tm[5]+1900, $tm[4]+1, $tm[3]), 'auto read from '. $meth_d);
		$c+=2;
	}
	return $c;
}

sub read_picture {
	my $tag = shift;
	my $testoptions = shift;
	my $c = 0;
	return 0 if (! $testoptions->{picture_read});
	ok($tag->picture_exists, 'Picture Exists');
	$c+=2;
	if ($testoptions->{picture_sha1}) {
		my $sha1 = Digest::SHA1->new();
		$sha1->add($tag->picture->{_Data});
		cmp_ok($sha1->hexdigest, 'eq', $testoptions->{picture_sha1}, 'digest of picture matches during read');
		$c++;
	}
}

sub write_picture {
	my $tag = shift;
	my $testoptions = shift;
	my $c = 0;
	return 0 if (! $testoptions->{picture_file});
	ok($tag->picture_filename($testoptions->{picture_file}), 'add picture');
	ok($tag->picture_exists, 'Picture Exists after write');
	$c+=2;
	if ($testoptions->{picture_sha1}) {
		my $sha1 = Digest::SHA1->new();
		$sha1->add($tag->picture->{_Data});
		cmp_ok($sha1->hexdigest, 'eq', $testoptions->{picture_sha1}, 'digest of picture matches after write');
		$c++;
	}
	return $c;
}

sub filetest {
    my $file        = shift;
    my $filetest    = shift;
    my $tagoptions  = shift;
    my $testoptions = shift;
	my $c = 0;

  SKIP: {
        skip ("File: $file does not exists", $testoptions->{count} || 1) if (! -f $file);
        return unless (-f $file);
        copy($file, $filetest);

        my $tag = create_tag($filetest,$tagoptions,$testoptions);
		$c+=3;
        die unless $tag;


		read_tag($tag,$testoptions);
		if ($testoptions->{picture_in}) {
			ok($tag->picture_exists, 'Picture should exists');
		}
		else {
			ok(! $tag->picture_exists, 'Picture should not exist');
		}
		$c++;

		if ($testoptions->{skip_write_tests}) {
			$tag->close();
			$tag = undef;
		}
		else {
			$c+= random_write($tag,$testoptions);
			$c+= random_write_num($tag,$testoptions);
			$c+= random_write_date($tag,$testoptions);
			$c+= write_picture($tag,$testoptions);
			ok($tag->set_tag, 'set_tag: ' . $filetest);
			$c++;
			$tag->close();
			$tag = undef;
			my $tag2 = create_tag($filetest,$tagoptions,$testoptions);
			$c+=3;
			$c+= random_read($tag2,$testoptions);
			$c+= random_read_num($tag2,$testoptions);
			$c+= random_read_date($tag2,$testoptions);
			$c+= read_picture($tag2,$testoptions);
			$tag2->close();
		}
		unlink($filetest);
		return $c;
    }
}


1;
__END__

=pod

=head1 NAME

Music::Tag::Test - Routines to test Music::Tag plugins

=head1 SYNOPSIS

    #!/usr/bin/perl -w
    use strict;
    use Test::More tests => 62;
    use Music::Tag::Test;

    BEGIN { use_ok('Music::Tag') }
    my $c = filetest("t/elise.mp3", "t/elisetest.mp3", {},{
        values_in => {
            artist =>, "Beethoven", 
            album => "GPL",
            title 1=> "Elise",
            sha1 => '39cd05447fa9ab6d6db08f41a78ac8628874c37e',
        },
        skip_write_tests => 0,
        random_write => [ qw(title artist album) ], 
        random_write_date => [ qw(releaseepoch) ],
        random_write_num => [ qw(track) ],
        picture_in => 0,
        picture_file => 'beethoven.jpg',
        picture_sha1 => 'b2bf4b2f71bf01e12473dd0ebe295777127589f4',
        picture_read => 1,
        count => 60,
        plugin => 'MP3'
    });

=head1 DESCRIPTION

Music::Tag::Test provides routines to test Music::Tag plugins

=head1 SUBROUTINES

=over

=item B<filetest()>

Takes a hashref of options to perform multiple test on a file.

=item B<create_tag>

Create a tag for given options.

=item B<read_tag()>

Perform read test

=item B<random_write()>, B<random_write_num()>, B<random_write_date()>

Write random data to tags

=item B<random_read()>, B<random_read_num()>, B<random_read_date()>

Read back data written.

=item B<read_picture>

Read a picture

=item B<write_picture>

Write a picture

=back

=head1 BUGS

Unknown

=head1 SEE ALSO

L<Music::Tag>

=head1 AUTHOR 

Edward Allen III <ealleniii _at_ cpan _dot_ org>

=head1 LICENSE

This program is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either:

    a) the GNU General Public License as published by the Free
    Software Foundation; either version 1, or (at your option) any
    later version, or

    b) the "Artistic License" which comes with Perl.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See either
the GNU General Public License or the Artistic License for more details.

You should have received a copy of the Artistic License with this
Kit, in the file named "Artistic".  If not, I'll be glad to provide one.

You should also have received a copy of the GNU General Public License
along with this program in the file named "Copying". If not, write to the
Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
Boston, MA 02110-1301, USA or visit their web page on the Internet at
http://www.gnu.org/copyleft/gpl.html.


=head1 COPYRIGHT

Copyright (c) 2007,2010 Edward Allen III. Some rights reserved.




lib/Music/Tag/Option.pm
package Music::Tag::Option;
use strict;
use warnings;
our $VERSION = .40_01;
use base qw(Music::Tag::Generic);

# Copyright (c) 2006 Edward Allen III. Some rights reserved.

#
# You may distribute under the terms of either the GNU General Public
# License or the Artistic License, as specified in the README file.
#



sub set_tag {
    my $self = shift;
    my $okmethods = { map { lc($_) => 1 } @{ $self->info->datamethods } };
    while ( my ( $k, $v ) = each %{ $self->options } ) {
        if ( ( defined $v ) and ( $okmethods->{ lc($k) } ) ) {
            my $method = lc($k);
            $self->info->$method($v);
            $self->tagchange($method);
        }
    }
}

sub get_tag { set_tag(@_); }

1;
__END__
=pod

=head1 NAME

Music::Tag::Option - Plugin module for Music::Tag to set tags via tag optons 

=head1 SYNOPSIS

	use Music::Tag

	my $filename = "/var/lib/music/artist/album/track.flac";

	my $info = Music::Tag->new($filename, { quiet => 1 }, "ogg");

	$info->add_plugin(option, { artist => "Sarah Slean" });

	$info->get_info();
	   
	print "Artist is ", $info->artist;

	#Outputs "Artist is Sarah Slean"

=head1 DESCRIPTION

Music::Tag::Option is a plugin to set tags via the plugin option.

=head1 REQUIRED VALUES

None.

=head1 SET VALUES

=over 4

=item Any value you would like can be set this way.

=back

=head1 OPTIONS

Any tag accepted by L<Music::Tag>.

=head1 METHODS

=over

=item default_options

Returns the default options for the plugin.  

=item set_tag

Sets the info in the Music::Tag file to info from options.

=item get_tag

Same as set_tag.

=back

=head1 BUGS

No known additional bugs provided by this Module.

=head1 SEE ALSO

L<Music::Tag>

=head1 AUTHOR 

Edward Allen III <ealleniii _at_ cpan _dot_ org>

=head1 LICENSE

This program is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either:

    a) the GNU General Public License as published by the Free
    Software Foundation; either version 1, or (at your option) any
    later version, or

    b) the "Artistic License" which comes with Perl.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See either
the GNU General Public License or the Artistic License for more details.

You should have received a copy of the Artistic License with this
Kit, in the file named "Artistic".  If not, I'll be glad to provide one.

You should also have received a copy of the GNU General Public License
along with this program in the file named "Copying". If not, write to the
Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
Boston, MA 02110-1301, USA or visit their web page on the Internet at
http://www.gnu.org/copyleft/gpl.html.


=head1 COPYRIGHT

Copyright (c) 2007,2010 Edward Allen III. Some rights reserved.



lib/Music/Tag/Generic.pm
package Music::Tag::Generic;
use strict;
use warnings;
our $VERSION = .40_01;

# Copyright (c) 2007,2008,2009,2010 Edward Allen III. Some rights reserved.

#
# You may distribute under the terms of either the GNU General Public
# License or the Artistic License, as specified in the README file.
#

use Encode;
use vars qw($AUTOLOAD);
use Scalar::Util qw(weaken);
use Carp;

sub new {
    my $class   = shift;
    my $parent  = shift;
    my $options = shift || {};
    my $self    = {};
    bless $self, $class;
    $self->info($parent);
    $self->options($options);
    return $self;
}

sub info {
    my $self = shift;
    my $val  = shift;
    if ( defined $val && ref $val ) {
        $self->{info} = $val;
        weaken $self->{info};
    }
    return $self->{info};
}

sub required_values {
}

sub set_values {
}

sub saved_values {
}

sub get_tag {
}

sub set_tag {
}

sub strip_tag {
}

sub close {
}

sub tagchange {
    my $self = shift;
    my $tag  = lc(shift);
    my $to   = shift || $self->info->$tag || "";
    $self->status( $self->info->_tenprint( $tag, 'bold blue', 15 ) . '"' . $to . '"' );
    return $self->info->changed(1);
}

sub simplify {
    my $self = shift;
    my $text = shift;
    chomp $text;
    return $text unless $text;

    if ( $self->options->{Unaccent} ) {
        $text = Text::Unaccent::PurePerl::unac_string($text);
    }

    $text = lc($text);

    $text =~ s/\[[^\]]+\]//g;
    $text =~ s/[\s_]/ /g;

    if ( length($text) > 5 ) {
        $text =~ s/\bthe\s//g;
        $text =~ s/\ba\s//g;
        $text =~ s/\ban\s//g;
        $text =~ s/\band\s//g;
        $text =~ s/\ble\s//g;
        $text =~ s/\bles\s//g;
        $text =~ s/\bla\s//g;
        $text =~ s/\bde\s//g;
    }
    if ( $self->options->{Inflect} ) {
        $text =~ s/(\.?\d+\,?\d*\.?\d*)/Lingua::EN::Inflect::NUMWORDS($1)/egxms;
    }
    else {
        $text =~ s/\b10\s/ten /g;
        $text =~ s/\b9\s/nine /g;
        $text =~ s/\b8\s/eight /g;
        $text =~ s/\b7\s/seven /g;
        $text =~ s/\b6\s/six /g;
        $text =~ s/\b5\s/five /g;
        $text =~ s/\b4\s/four /g;
        $text =~ s/\b3\s/three /g;
        $text =~ s/\b2\s/two /g;
        $text =~ s/\b1\s/one /g;
    }

    $text =~ s/\sii\b/two/g;
    $text =~ s/\siii\b/three/g;
    $text =~ s/\siv\b/four/g;
    $text =~ s/\sv\b/five/g;
    $text =~ s/\svi\b/six/g;
    $text =~ s/\svii\b/seven/g;
    $text =~ s/\sviii\b/eight/g;

    # Don't translate IX because of a soft spot in my heart for the technologically rich planet.

    $text =~ s/[^a-z0-9]//g;
    return $text;
}

sub simple_compare {
    my $self            = shift;
    my $a               = shift;
    my $b               = shift;
    my $similar_percent = shift;
    my $crop_percent    = shift;

    my $sa = $self->simplify($a);
    my $sb = $self->simplify($b);
    if ( $sa eq $sb ) {
        return 1;
    }

    return unless ( $similar_percent || $crop_percent );

    my $la  = length($sa);
    my $lb  = length($sb);
    my $max = ( $la < $lb ) ? $lb : $la;
    my $min = ( $la < $lb ) ? $la : $lb;

    return unless ( $min and $max );

    my $dist = undef;
    if ( $self->options->{LevenshteinXS} ) {
        $dist = Text::LevenshteinXS::distance( $sa, $sb );
    }
    elsif ( $self->options->{Levenshtein} ) {
        $dist = Text::Levenshtein::distance( $sa, $sb );
    }
    unless ($crop_percent) {
        $crop_percent = $similar_percent * ( 2 / 3 );
    }

    if ( ( defined $dist ) && ( ( ( $min - $dist ) / $min ) >= $similar_percent ) ) {
        return -1;
    }

    if ( $min < 10 ) {
        return 0;
    }
    if ( ( ( ( 2 * $min ) - $max ) / $min ) <= $crop_percent ) {
        return 0;
    }
    if ( substr( $sa, 0, $min ) eq substr( $sb, 0, $min ) ) {
        return -1;
    }
    return 0;
}

sub status {
    my $self = shift;
    $self->info->status( ref($self), @_ );
	return;
}

sub error {
    my $self = shift;
    carp( ref($self), " ", @_ );
	return;
}

sub changed {
    my $self = shift;
    return $self->info->changed(@_);
}

sub wav_out {
    my $self = shift;
    my $fh   = shift;
    if ( $self->options->{wav_out_system} ) {
        my @sys = ();
        foreach ( @{ $self->options->{wav_out_system} } ) {
            my $a = $_;
            $a =~ s/\[FILENAME\]/$self->info->filename()/ge;
            push @sys, $a;
        }
        $self->status( 0, "Executing ", join( " ", @sys ) );
		my $in;
        if ( open( $in, '-|', @sys ) ) {
            binmode $in;
            binmode $fh;
            my $buffer = "";
            while ( my $count = sysread( $in, $buffer, 1024 ) ) {
                my $wrote = 0;
                while ( $wrote < $count ) {
                    $wrote += syswrite( $fh, $buffer, ( $count - $wrote ), $wrote );
                }
                $buffer = "";
            }
            return 1;
        }
		CORE::close($in);
        return 0;
    }
    return;
}

sub options {
    my $self = shift;
    unless ( exists $self->{_options} ) {
        $self->{_options} = Config::Options->new( $self->default_options );
    }
    return $self->{_options}->options(@_);
}

sub default_options { return {} }

sub DESTROY {
    my $self = shift;
    if ( exists $self->{info} ) {
		delete $self->{info};
    }
	return;
}

1;

__END__
=pod

=head1 NAME

Music::Tag::Generic - Parent Class for Music::Tag objects

=head1 SYNOPSIS

    package Music::Tag::SuperMusic;

	use base qw(Music::Tag::Generic);

	sub set_tag {
		my $self = shift;
		$self->info->artist($self->info->artist . " is super");
		return $self->info;
	}

	1;

=head1 DESCRIPTION

Base class.  See L<Music::Tag|Music::Tag>.

=pod

=head1 PLUGINS

All plugins should set @ISA to include Music::Tag::Generic and contain one or more of the following methods:

=over 4

=item B<new()>

Set in template. If you override, it should take as options a reference to a Music::Tag object and an href of options. 

=pod

=item B<info()>

Should return a reference to the associated Music::Tag object. If passed an object, should set the associated Music::Tag object to it.

=item B<get_tag()>

Populates the data in the Music::Tag object.

=item B<set_tag()>

Optional method to save info.

=item B<required_values()>

Optional method returns a list of required data values required for L<Music::Tag::Generic/get_tag()>.

=item B<set_values()>

Optional method (for now) returns a list of data values that can be set with L<Music::Tag::Generic/get_tag()>.

=item B<saved_values()>

Optional method returns a list of data values that can be saved with L<Music::Tag::Generic/set_tag()>.

=item B<strip_tag()>

Optional method to remove info. 

=item B<close()>

Optional method to close open file handles.

=item B<tagchange()>

Inherited method that can be called to announce a data-value change from what is read on file. Used by secondary plugins like Amazon, MusicBrainz, and File.  This is preferred to using C<<$self->info->changed(1)>>.

=item B<simplify()>

A useful method for simplifying artist names and titles. Takes a string, and returns a sting with no whitespace.  Also removes accents (if Text::Unaccent::PurePerl is available) and converts numbers like 1,2,3 as words to one, two, three... (English is used here.  Let me know if it would be helpful to change this. I do not change words to numbers because I prefer sorting "5 Star" under f).  Removes known articles, such as a, the, an, le les, de if they are not at the end of a string. 

=item B<simple_compare>($a, $b, $required_percent)

Returns 1 on match, 0 on no match, and -1 on approximate match.   $required_percent is
a value from 0...1 which is the percentage of similarity required for match.  

=item B<status()>

Inherited method to print a pretty status message. If first argument is a number, assumes this is required
verbosity. 

=item B<error()>

Inherited method to print an error message.

=item B<changed()>

Same as $self->info->changed().  Please use L<tagchange> method instead.

=item B<wav_out()>

If plugin is for a media tag, return stream of wav to filehandle $fh. 

Return True on success, False on failure, undef if not supported.

=item B<options()>

Returns a hashref of options (or sets options, just like Music::Tag method).

=pod

=item B<default_options>

Method should return default options.

=back

=head1 SEE ALSO

L<Music::Tag|Music::Tag>

=head1 AUTHOR 

Edward Allen III <ealleniii _at_ cpan _dot_ org>

=head1 COPYRIGHT

Copyright (c) 2007,2008 Edward Allen III. Some rights reserved.

=head1 LICENSE

This program is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.



lib/Music/Tag/Auto.pm
package Music::Tag::Auto;
use strict;
use warnings;
our $VERSION = .40_01;
use base qw(Music::Tag::Generic);

# Copyright (c) 2006 Edward Allen III. Some rights reserved.

#
# You may distribute under the terms of either the GNU General Public
# License or the Artistic License, as specified in the README file.
#

sub default_options {
	{
		autoplugin => {
			mp3	=> "MP3",
			m4a => "M4A",
			m4p => "M4A",
			mp4 => "M4A",
			m4b => "M4A",
			'3gp' => "M4A",
			ogg => "OGG",
			flac => "FLAC"
		}
	}
}

sub new {
    my $class   = shift;
    my $parent  = shift;
    my $options = shift || {};
    my $self    = {};
    bless $self, $class;
    $self->info($parent);
    $self->options($options);
    my $plugin   = "";

    if ( $self->info->filename =~ /\.([^\.]*)$/ ) {
		if (exists $self->options->{autoplugin}->{lc($1)}) {
		   $plugin = $self->options->{autoplugin}->{lc($1)}; 
		}
    }
	if (($plugin) && ($self->info->available_plugins($plugin))) {
		unless ( $plugin =~ /::/ ) {
			$plugin = "Music::Tag::" . $plugin;
		}
		$self->status(1, "Auto loading plugin: $plugin");
	my $type = lc($plugin);
	$type =~ s/^.*:://;
        $self->status(2, "Adding filetype: $plugin");
        $self->info->filetype($type);
		if($self->info->_has_module($plugin)) {
			return $plugin->new( $self->info, $self->options );
		}
    }
    else {
        $self->error("Sorry, I can't find a plugin for ", $self->info->filename);
        return undef;
    }
}

1;
__END__
=pod

=head1 NAME

Music::Tag::Auto - Plugin module for Music::Tag to load other plugins by file extension.

=head1 SYNOPSIS

	use Music::Tag

	my $filename = "/var/lib/music/artist/album/track.flac";

	my $info = Music::Tag->new($filename, { quiet => 1 }, "Auto");

	$info->get_info();
	print "Artist is ", $info->artist;


=head1 DESCRIPTION

Music::Tag::Auto is loaded automatically in Music::Tag .3 and newer to load other plugins.

=head1 REQUIRED VALUES

None.

=head1 SET VALUES

None.

=head1 OPTIONS

=over 4

=item B<autoplugin>

Option is a hash reference.  Reference maps file extensions to plugins. Default is: 

    {   mp3	  => "MP3",
        m4a   => "M4A",
        m4p   => "M4A",
        mp4   => "M4A",
        m4b   => "M4A",
        '3gp' => "M4A",
        ogg   => "OGG",
        flac  => "FLAC"   }

=back

=head1 METHODS

=over

=item new($parent, $options)

Returns a Music::Tag object based on file extension, if available.  Otherwise returns undef. 

=item default_options

Returns the default options for the plugin.  

=item set_tag

Not defined for this plugin.

=item get_tag

Not defined for this plugin.

=back

=head1 BUGS

No known additional bugs provided by this Module.

=head1 SEE ALSO

L<Music::Tag>, L<Music::Tag::FLAC>, L<Music::Tag::Lyrics>, L<Music::Tag::M4A>, L<Music::Tag::MP3>, L<Music::Tag::OGG>

=head1 AUTHOR 

Edward Allen III <ealleniii _at_ cpan _dot_ org>

=head1 LICENSE

This program is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either:

    a) the GNU General Public License as published by the Free
    Software Foundation; either version 1, or (at your option) any
    later version, or

    b) the "Artistic License" which comes with Perl.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See either
the GNU General Public License or the Artistic License for more details.

You should have received a copy of the Artistic License with this
Kit, in the file named "Artistic".  If not, I'll be glad to provide one.

You should also have received a copy of the GNU General Public License
along with this program in the file named "Copying". If not, write to the
Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
Boston, MA 02110-1301, USA or visit their web page on the Internet at
http://www.gnu.org/copyleft/gpl.html.


=head1 COPYRIGHT

Copyright (c) 2007 Edward Allen III. Some rights reserved.



